### ISSUE TITLE
Set attribute value only on first object creation

### ISSUE THREAD

**Issue by Visya** [pynamodb/PynamoDB#810]
Hi!

**TL;DR**
I'm wondering if there is a way to know in `__init__` if the object is newly created in the table or if an existing object is retrieved from the DynamoDB.
Or is there a way to update an attribute only for newly created objects?

---

I have an abstract class `AbstractPost` that has several children `ApprovedPost`, `RejectedPost`, `InboxPost`. Whenever a post is created in a collection, I want to set a specific attribute to `datetime.now()`, regardless if the attribute was supplied to the `__init__` (which happens when I copy post to another collection).

Here is a simplified version of my code:

``` python
class AbstractPost(Model):
  url = UnicodeAttribute(null=True)
  stateChangedAt = UTCDateTimeAttribute(default_for_new=datetime.now) # Should be always datetime.now for new objects

  class Meta(object):
    abstract = True

  def __init__(self, *args: Any, **kwargs: Any) -> None:
    super(AbstractPost, self).__init__(*args, **kwargs)
    self.stateChangedAt = datetime.now()

class ApprovedPost(AbstractPost):
  pass

class RejectedPost(AbstractPort):
  pass

rejected_post = RejectedPost(url='http://my.url') # Here stateChangedAt is set to now
approved_post = ApprovedPost(rejected_post.__dict__) # Here stateChangedAt should be set to the new now timestamp
```

As you can see, I tried to reset `stateChangedAt` to now in the `__init__` function. The problem now is that when I query DynamoDB for posts, `__init__` is called and `stateChangedAt` is reset to now in query results.

Is there a way to distinguish between when the new post is created in the collection and when we retrieve an existing post? Or is there any way to define an attribute that is set only on object creation?

**Comment by ikonst** [comment]
We've actually been passing an undocumented argument `_user_instantiated` to `__init__` for this same purpose:
https://github.com/pynamodb/PynamoDB/blob/8d58d2e20245a7d0f3d2fb57c9b9760c40ac77c4/pynamodb/attributes.py#L254

You can use it as long as you don't mind using an undocumented feature that might change later:
```
def __init__(self, *, _user_initiated: bool, **kwargs: Any) -> None:
	super(AbstractPost, self).__init__(_user_initiated=_user_initiated, **kwargs)
	if _user_initiated:
		self.stateChangedAt = datetime.now()
```
or:
```
def __init__(self, *, _user_initiated: bool, **kwargs: Any) -> None:
    if _user_initiated:
	    kwargs['setChangedAt'] = datetime.now()
	super(AbstractPost, self).__init__(_user_initiated=_user_initiated, **kwargs)
```
or even:
```
def __init__(self, *, _user_initiated: bool, **kwargs: Any) -> None:
    if _user_initiated:
		if 'setChangedAt' in kwargs:
		   raise ValueError("you're initializing setChangedAt, is it a bug?")
	    kwargs['setChangedAt'] = datetime.now()
	super(AbstractPost, self).__init__(_user_initiated=_user_initiated, **kwargs)
```

**Comment by Visya** [comment]
Thank you!

In the end I did something like this, because it wouldn't let me destructure `_user_instantiated` directly in the constructor agruments.
```
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        user_instantiated = kwargs.get("_user_instantiated", True)

        if user_instantiated:
            kwargs["stateChangedAt"] = datetime.now()

        super(AbstractPost, self).__init__(*args, **kwargs)
```
